### PR TITLE
Add AMReX

### PR DESCRIPTION
**Pull request by ax3l**
Add AMReX: A software framework for massively parallel, block-structured adaptive mesh refinement (AMR) applications.

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] ~If static libraries are linked in, the license of the static library is packaged.~
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by ax3l** [comment]
> GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.

Ping @atmyers @asalmgren @WeiqunZhang

**Comment by atmyers** [comment]
I'm willing to be listed as a maintainer.

**Comment by WeiqunZhang** [comment]
I am willing to be listed as a maintainer.

**Comment by ax3l** [comment]
@conda-forge/help-c-cpp ready for review / merge :)

File: recipes/amrex/bld.bat [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    -DCMAKE_CXX_COMPILER=clang-cl   ^
+    -DCMAKE_INSTALL_LIBDIR=lib      ^
+    -DCMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX% ^
+    -DCMAKE_LINKER=lld-link         ^
+    -DCMAKE_NM=llvm-nm              ^
+    -DCMAKE_VERBOSE_MAKEFILE=ON
+if errorlevel 1 exit 1
+
+:: build
+cmake --build build --config Release --parallel 2
```
Comment (isuruf) [inline recipes/amrex/bld.bat:44]:
```suggestion
cmake --build build --config Release --parallel %CPU_COUNT%
```

Comment (ax3l) [reply recipes/amrex/bld.bat:44]: Ah, I was not sure anymore if that variable was set on Windows, too. Thanks!

**Comment by isuruf** [comment]
Following filenames are super generic names. Are these needed and if they are can they get a AMReX prefix?
```
C:/bld/amrex_1697583800068/_h_env/Library/bin/fboxinfo.exe
C:/bld/amrex_1697583800068/_h_env/Library/bin/fcompare.exe
C:/bld/amrex_1697583800068/_h_env/Library/bin/fextract.exe
C:/bld/amrex_1697583800068/_h_env/Library/bin/fextrema.exe
C:/bld/amrex_1697583800068/_h_env/Library/bin/fgradient.exe
C:/bld/amrex_1697583800068/_h_env/Library/bin/fnan.exe
C:/bld/amrex_1697583800068/_h_env/Library/bin/fsnapshot.exe
C:/bld/amrex_1697583800068/_h_env/Library/bin/ftime.exe
C:/bld/amrex_1697583800068/_h_env/Library/bin/fvarnames.exe
C:/bld/amrex_1697583800068/_h_env/Library/bin/fvolumesum.exe
C:/bld/amrex_1697583800068/_h_env/Library/Tools/CMake/FindFilesystem.cmake
C:/bld/amrex_1697583800068/_h_env/Library/Tools/CMake/FindHYPRE.cmake
C:/bld/amrex_1697583800068/_h_env/Library/Tools/CMake/FindPETSc.cmake
C:/bld/amrex_1697583800068/_h_env/Library/Tools/C_scripts
C:/bld/amrex_1697583800068/_h_env/Library/Tools/C_scripts/describe_sources.py
C:/bld/amrex_1697583800068/_h_env/Library/Tools/C_scripts/gatherbuildtime.py
C:/bld/amrex_1697583800068/_h_env/Library/Tools/C_scripts/makebuildinfo_C.py
C:/bld/amrex_1697583800068/_h_env/Library/Tools/C_scripts/mkdep
C:/bld/amrex_1697583800068/_h_env/Library/Tools/C_scripts/mmclt.py
C:/bld/amrex_1697583800068/_h_env/Library/Tools/C_scripts/strip72
C:/bld/amrex_1697583800068/_h_env/Library/Tools/typechecker
C:/bld/amrex_1697583800068/_h_env/Library/Tools/typechecker/typechecker.py
```

**Comment by ax3l** [comment]
Thanks @isuruf, we are discussing to either skip the tools for now or adding suffix/prefixes.

I will move the `Tool/` directory also to something properly aligned, e.g., `shared/amrex/`.

File: recipes/amrex/bld.bat [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    -DAMReX_GPU_BACKEND=NONE        ^
+    -DAMReX_GPU_RDC=OFF             ^
+    -DAMReX_HDF5=OFF                ^
+    -DAMReX_HYPRE=OFF               ^
+    -DAMReX_IPO=OFF                 ^
+    -DAMReX_MPI=OFF                 ^
+    -DAMReX_MPI_THREAD_MULTIPLE=OFF ^
+    -DAMReX_OMP=ON                  ^
+    -DAMReX_PARTICLES=ON            ^
+    -DAMReX_PLOTFILE_TOOLS=ON       ^
```
Comment (ax3l) [inline recipes/amrex/bld.bat:25]: To skip tools until we add unique suffixes/prefixes in future releases.
```suggestion
    -DAMReX_PLOTFILE_TOOLS=OFF      ^
```

File: recipes/amrex/build.sh [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    -DAMReX_GPU_BACKEND=NONE          \
+    -DAMReX_GPU_RDC=OFF               \
+    -DAMReX_HDF5=OFF                  \
+    -DAMReX_HYPRE=OFF                 \
+    -DAMReX_IPO=OFF                   \
+    -DAMReX_MPI=OFF                   \
+    -DAMReX_MPI_THREAD_MULTIPLE=OFF   \
+    -DAMReX_OMP=ON                    \
+    -DAMReX_PARTICLES=ON              \
+    -DAMReX_PLOTFILE_TOOLS=ON         \
```
Comment (ax3l) [inline recipes/amrex/build.sh:35]: To skip tools until we add unique suffixes/prefixes in future releases.
```suggestion
    -DAMReX_PLOTFILE_TOOLS=OFF        \
```

**Comment by ax3l** [comment]
OSX somehow had an environment resolution issue on `llvm-openmp` after everything else passed. Seems unrelated, restarted CI now.

File: recipes/amrex/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    - ccache        # [unix]
+    - clang         # [win]
+    - cmake
+    - lld           # [win]
+    # TODO: add libxml2 dep in lld-feedstock
+    - libxml2       # [win]
+    - make          # [unix]
+    - ninja         # [win]
+    - libgomp       # [linux]
+    - llvm-openmp   # [osx or win]
```
Comment (ax3l) [inline recipes/amrex/meta.yaml:32]: @isuruf since OpenMP has a platform-specific runtime, shouldn't it be in `host:`?

**Comment by ax3l** [comment]
@isuruf thank you for the guidance - all points addressed :+1:

**Comment by ax3l** [comment]
Oh dang, the Windows modules for CMake are still a bit all over the place:
```
-- Installing: C:/bld/amrex_1697676564380/_h_env/Library/lib/cmake/AMReX/AMReXTargets.cmake
-- Installing: C:/bld/amrex_1697676564380/_h_env/Library/lib/cmake/AMReX/AMReXTargets-release.cmake
-- Installing: C:/bld/amrex_1697676564380/_h_env/Library/cmake/Modules/FindFilesystem.cmake
-- Installing: C:/bld/amrex_1697676564380/_h_env/Library/cmake/Modules/FindHYPRE.cmake
-- Installing: C:/bld/amrex_1697676564380/_h_env/Library/cmake/Modules/FindPETSc.cmake
```

**Comment by ax3l** [comment]
Fixed:
- `Modules` -> `AMReXCMakeModules`
- `AMReXTargets*.cmake` to correct place on Windows

**Comment by ax3l** [comment]
Ok, now we are good.

Windows:
```
...
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXConfig.cmake
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXConfigVersion.cmake
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXTargets.cmake
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXTargets-release.cmake
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/AMReXTargetHelpers.cmake
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/AMReXThirdPartyLibraries.cmake
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/AMReXTypecheck.cmake
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/AMReXUtils.cmake
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/AMReX_buildInfo.cpp.in
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/AMReX_Config.cmake
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/AMReX_Config.H.in
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/AMReX_Config_ND.H.in
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/AMReX_SetupCUDA.cmake
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/AMReX_ThirdPartyProfilers.cmake
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/AMReX_Version.H.in
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/FindFilesystem.cmake
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/FindHYPRE.cmake
-- Installing: C:/bld/amrex_1697681384400/_h_env/Library/cmake/AMReXCMakeModules/FindPETSc.cmake
```

**Comment by ax3l** [comment]
@isuruf ready for merge now 👍

**Comment by ax3l** [comment]
I posted a follow-up for `pyAMReX` in https://github.com/conda-forge/staged-recipes/pull/24298

**Pull request merged**